### PR TITLE
Refactor: factorise plugins app implementation

### PR DESCRIPTION
**Pull request by sfc-gh-pczajka**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [ ] I've added or updated automated unit tests to verify correctness of my new code.
   * [ ] I've added or updated integration tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [x] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.

### Changes description
Refactor: change SnowTyper into Typer instance fanctory.
Only developer-facing change: we need to call `SnowTyper.create_app()` to get an actual Typer instance.

The motivation behind the change is an issue https://github.com/snowflakedb/snowflake-cli/issues/813, which was caused by reading `config.toml` before error handling was set up

### CODE REVIEW

File: src/snowflake/cli/plugins/connection/commands.py [hunk 1]
```diff
@@ -55,7 +64,7 @@ def _mask_password(connection_params: dict):
     return connection_params
 
 
-@app.command(name="list")
+@app_creator.command(name="list")
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/plugins/connection/commands.py:67]: Using `app` will reduce a lot of changes, I don't think we need to to the rename

File: src/snowflake/cli/plugins/connection/commands.py [hunk 2]
```diff
@@ -32,10 +32,19 @@
...
-)
+
+class ConnectionAppCreator(SnowTyperCreator):
+    def create_app(self) -> SnowTyper:
+        app = SnowTyper(
+            name="connection",
+            help="Manages connections to Snowflake.",
+        )
+        self.register_commands(app)
+        return app
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/plugins/connection/commands.py:43]: Can't we simplify if to something like
```py
app: SnowTyperCreator = SnowTyperCreator.create_app(SnowTyper(
            name="connection",
            help="Manages connections to Snowflake.",
        ))
```
do we really need a custom class for every command group?

File: src/snowflake/cli/api/commands/snow_typer.py [hunk 3]
```diff
@@ -111,3 +112,57 @@ def post_execute():
...
+            self.commands_to_register.append(
+                SnowTyperCommandData(command, args=args, kwargs=kwargs)
+            )
+            return command
+
+        return decorator
+
+    def register_commands(self, app: SnowTyper) -> None:
+        for command in self.commands_to_register:
+            app.command(*command.args, **command.kwargs)(command.func)
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/commands/snow_typer.py:168]: Why we cannot add this logic to `SnowTyper`? What's reasoning behind adding this additional layer?

File: src/snowflake/cli/plugins/connection/commands.py (hunk 2 above)
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/plugins/connection/commands.py:43]: I could pass `**kwargs` from `create_app` to the constructor, but then I'd have to modify `add_init_command`-like procedures to work with `SnowTyperCreator` - I will check

File: src/snowflake/cli/api/commands/snow_typer.py (hunk 3 above)
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/api/commands/snow_typer.py:168]: We want to postpone calling `__init__` until error handling would be set up. The problem behind the issue was calling feature flags (which read config.toml) to determine constructor arguments (`hidden=FeatureFlag.is_disabled`)

File: src/snowflake/cli/api/commands/snow_typer.py [hunk 4]
```diff
@@ -111,3 +112,73 @@ def post_execute():
...
+
+
+@dataclasses.dataclass
+class SnowTyperCommandData:
+    func: Callable
+    args: Tuple[Any, ...]
+    kwargs: Dict[str, Any]
+
+
+class SnowTyper:
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/commands/snow_typer.py:124]: Should we name it `SnowTyperFactory`?

File: src/snowflake/cli/api/commands/snow_typer.py [hunk 5]
```diff
@@ -111,3 +112,73 @@ def post_execute():
...
+    creates actual SnowTyperInstance instance.
+    """
+
+    def __init__(
+        self,
+        /,
+        name: Optional[str] = None,
+        help: Optional[str] = None,  # noqa: A002
+        short_help: Optional[str] = None,
+        is_hidden: Optional[Callable[[], bool]] = None,
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/commands/snow_typer.py:136]:
```suggestion
        is_enabled: Optional[Callable[[], bool]] = None,
```
To be consistent with commands option

File: src/snowflake/cli/plugins/connection/commands.py (hunk 1 above)
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/plugins/connection/commands.py]: changed

File: src/snowflake/cli/plugins/connection/commands.py (hunk 2 above)
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/plugins/connection/commands.py]: changed as discussed

File: src/snowflake/cli/api/commands/snow_typer.py (hunk 3 above)
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/api/commands/snow_typer.py]: changed as discussed

File: src/snowflake/cli/api/commands/snow_typer.py (hunk 5 above)
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/api/commands/snow_typer.py:136]: discussed on Slack

**Review by sfc-gh-pjob — APPROVED** [review]

File: src/snowflake/cli/api/commands/snow_typer.py [hunk 6]
```diff
@@ -111,3 +112,73 @@ def post_execute():
 
         log.debug("Executing command post execution callback")
         flush_telemetry()
+
+
+@dataclasses.dataclass
+class SnowTyperCommandData:
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/commands/snow_typer.py:118]: Let's add comment what's the purpose of the class

File: src/snowflake/cli/api/commands/snow_typer.py (hunk 4 above)
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/api/commands/snow_typer.py]: renamed:
`SnowTyper` -> `SnowTyperFactory`
`SnowTyperInstance` -> `SnowTyper`
`.create_app()` -> `.create_instance()`

File: src/snowflake/cli/api/commands/snow_typer.py (hunk 6 above)
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/api/commands/snow_typer.py:118]: done!

File: src/snowflake/cli/plugins/nativeapp/commands.py
Comment (sfc-gh-bgoel) [inline src/snowflake/cli/plugins/nativeapp/commands.py:1]: + src/snowflake/cli/plugins/nativeapp/version/commands.py also needs these changes.

**Review by sfc-gh-yiwang — APPROVED** [review]

File: src/snowflake/cli/plugins/nativeapp/commands.py
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/plugins/nativeapp/commands.py:1]: It is already changed :)

File: src/snowflake/cli/plugins/object/__init__.py [hunk 7]
```diff
@@ -1,11 +0,0 @@
...
-from snowflake.cli.plugins.object.commands import app as show_app
-from snowflake.cli.plugins.stage.commands import app as stage_app
-
-app = SnowTyper(
-    name="object",
-    help="Manages Snowflake objects like warehouses and stages",
-)
-
-app.add_typer(stage_app)  # type: ignore
-app.add_typer(show_app)  # type: ignore
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/plugins/object/__init__.py:11]: why?

**Review by sfc-gh-turbaszek — DISMISSED** [review]

File: src/snowflake/cli/plugins/object/__init__.py (hunk 7 above)
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/plugins/object/__init__.py:11]: I made it consistent with other plugins (spcs / NADE), but I made the choice between `__init__.py` and `commands.py` arbitrarily - no strong opinions

File: tests_e2e/__snapshots__/test_error_handling.ambr [hunk 8]
```diff
@@ -0,0 +1,10 @@
+# serializer version: 1
+# name: test_corrupted_config_in_default_location
+  '''
+  ╭─ Error ──────────────────────────────────────────────────────────────────────╮
+  │ Configuration file seems to be corrupted. Key "demo" already exists. at line │
+  │ 2 col 18                                                                     │
+  ╰──────────────────────────────────────────────────────────────────────────────╯
+  
+  '''
+# ---
```
Comment (sfc-gh-turbaszek) [inline tests_e2e/__snapshots__/test_error_handling.ambr:10]: 🚀

**Review by sfc-gh-turbaszek — DISMISSED** [review]

**Review by sfc-gh-bgoel — APPROVED** [review]
lgtm from NADE

**Review by sfc-gh-bgoel — APPROVED** [review]
LGTM, acking the rebase.

**Review by sfc-gh-turbaszek — APPROVED** [review]

**Pull request merged**